src/librustc_mir/borrow_check/error_reporting.rs

```diff
@@ -518,6 +526,36 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         err.buffer(&mut self.errors_buffer);
     }
 
+    /// Returns a description of a place and an associated message for the purposes of conflicting
+    /// borrow diagnostics.
+    ///
+    /// If the borrow is of the field `b` of a union `u`, then the return value will be
+    /// `("u", " (via \`u.b\`)")`. Otherwise, for some variable `a`, the return value will be
+    /// `("a", "")`.
+    pub(super) fn describe_place_for_conflicting_borrow(
+        &self,
+        place: &Place<'tcx>,
+    ) -> (String, String) {
+        place.base_local()
```

Comment:
Hmm, this seems a bit special-cased. We could encounter a union that is not found in a local variable:

```rust
struct Foo {
    u: Bar
}

union Bar {
    x: u32, y: u32 
}
```

with an access to `foo.f`?

Similarly, this may trigger at times when it doesn't necessarily *need* to -- for example, if there is an example where we just access the same field `union.x` twice? (To avoid that, you would need to consider both places, I imagine)

Reply:
Pushed a commit that changes this logic so it can handle these cases.